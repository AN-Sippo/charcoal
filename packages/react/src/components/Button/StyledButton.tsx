import styled, { css } from 'styled-components'
import { variantToFont } from './lib/variantToFont'
import { variantToBackground } from './lib/variantToBackground'
import type { Size } from '.'

const horizontalPaddingSmall = css`
  padding-right: 16px;
  padding-left: 16px;
`

const horizontalPaddingMedium = css`
  padding-right: 24px;
  padding-left: 24px;
`

type StyledButtonProps = {
  $fullWidth: boolean
  $size: Size
  $background: ReturnType<typeof variantToBackground>
  $color: ReturnType<typeof variantToFont>
}

export const StyledButton = styled.button<StyledButtonProps>`
  appearance: none;
  background: transparent;
  padding: 0;
  border-style: none;
  outline: none;
  color: inherit;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  text-decoration: none;

  &:focus {
    outline: none;
  }

  width: ${(p) => (p.$fullWidth ? 'stretch' : 'min-content')};
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  border-radius: 999999px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;

  ${(p) => (p.$size === 'M' ? horizontalPaddingMedium : horizontalPaddingSmall)}
  color: var(--charcoal-${(p) => p.$color});
  background-color: var(--charcoal-${(p) => p.$background});
  transition: 0.2s color, 0.2s background-color, 0.2s box-shadow;

  &:not(:disabled):not([aria-disabled]),
  &[aria-disabled='false'] {
    &:active,
    &:focus,
    &:focus-visible {
      outline: none;
      box-shadow: 0 0 0 4px rgba(0, 150, 250, 0.32);
    }

    &:hover {
      color: var(--charcoal-${(p) => p.$color}-hover);
      background-color: var(--charcoal-${(p) => p.$background}-hover);
    }
    &:active {
      color: var(--charcoal-${(p) => p.$color}-press);
      background-color: var(--charcoal-${(p) => p.$background}-press);
    }
  }

  &:disabled,
  &[aria-disabled]:not([aria-disabled='false']) {
    cursor: default;
    opacity: 0.32;
  }
  height: ${(p) => (p.$size === 'M' ? 40 : 32)}px;
`
