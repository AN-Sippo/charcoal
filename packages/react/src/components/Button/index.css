.charcoal-button {
  appearance: none;
  background: transparent;
  box-sizing: border-box;
  padding: 0 24px;
  border-style: none;
  outline: none;
  text-rendering: inherit;
  letter-spacing: inherit;
  word-spacing: inherit;
  text-decoration: none;
  font: inherit;
  margin: 0;
  overflow: visible;
  text-transform: none;
  width: min-content;
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;
  border-radius: 999999px;
  font-size: 14px;
  line-height: 22px;
  font-weight: bold;
  color: var(--charcoal-color-text-secondary-default);
  background-color: var(--charcoal-color-container-secondary-default);
  transition: 0.2s color, 0.2s background-color, 0.2s box-shadow;
  height: 40px;
}

.charcoal-button:disabled {
  cursor: default;
  opacity: 0.32;
}

.charcoal-button:not(:disabled):focus-visible {
  outline: none;
  box-shadow: var(--charcoal-color-custom-focus-effect);
}

.charcoal-button:not(:disabled):hover {
  color: var(--charcoal-color-text-secondary-hover);
  background-color: var(--charcoal-color-container-secondary-hover);
}

.charcoal-button:not(:disabled):active,
.charcoal-button[data-active='true'] {
  color: var(--charcoal-color-text-secondary-press);
  background-color: var(--charcoal-color-container-secondary-press);
}

.charcoal-button[data-variant='Primary'] {
  color: var(--charcoal-color-text-on-primary-default);
  background-color: var(--charcoal-color-container-primary-default);
}

.charcoal-button[data-variant='Primary']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-primary-hover);
}

.charcoal-button[data-variant='Primary']:active:not(:disabled),
.charcoal-button[data-variant='Primary'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-primary-press);
}

.charcoal-button[data-variant='Overlay'] {
  color: var(--charcoal-color-text-on-primary-default);
  background-color: var(--charcoal-color-container-neutral-default);
}

.charcoal-button[data-variant='Overlay']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-neutral-hover);
}

.charcoal-button[data-variant='Overlay']:active:not(:disabled),
.charcoal-button[data-variant='Overlay'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-neutral-press);
}

.charcoal-button[data-variant='Navigation'] {
  color: var(--charcoal-color-text-on-h-u-d-default);
  background-color: var(--charcoal-color-container-h-u-d-default);
}

.charcoal-button[data-variant='Navigation']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-primary-hover);
  background-color: var(--charcoal-color-container-h-u-d-hover);
}

.charcoal-button[data-variant='Navigation']:active:not(:disabled),
.charcoal-button[data-variant='Navigation'][data-active='true'] {
  color: var(--charcoal-color-text-on-primary-press);
  background-color: var(--charcoal-color-container-h-u-d-press);
}

.charcoal-button[data-variant='Danger'] {
  color: var(--charcoal-color-text-on-negative-default);
  background-color: var(--charcoal-color-container-negative-default);
}
.charcoal-button[data-variant='Danger']:hover:not(:disabled) {
  color: var(--charcoal-color-text-on-negative-hover);
  background-color: var(--charcoal-color-container-negative-hover);
}

.charcoal-button[data-variant='Danger']:active:not(:disabled),
.charcoal-button[data-variant='Danger'][data-active='true'] {
  color: var(--charcoal-color-text-on-negative-press);
  background-color: var(--charcoal-color-container-negative-press);
}

.charcoal-button[data-size='S'] {
  padding: 0 16px;
  height: 32px;
}

.charcoal-button[data-full-width='true'] {
  width: 100%;
}
