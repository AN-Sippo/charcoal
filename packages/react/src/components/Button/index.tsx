import { forwardRef } from 'react'
import styled from 'styled-components'
import { unreachable } from '../../_lib'
import { theme } from '../../styled'
import Clickable, { ClickableElement, ClickableProps } from '../Clickable'

type Variant = 'Primary' | 'Default' | 'Overlay' | 'Danger' | 'Navigation'
type Size = 'S' | 'M'

interface StyledProps {
  /**
   * ボタンのスタイル
   */
  variant: Variant
  /**
   * ボタンのサイズ
   */
  size: Size
  /**
   * 幅を最大まで広げて描画
   */
  fullWidth: boolean
}

export type ButtonProps = Partial<StyledProps> & ClickableProps

const Button = forwardRef<ClickableElement, ButtonProps>(function Button(
  {
    children,
    variant = 'Default',
    size = 'M',
    fullWidth: fixed = false,
    disabled = false,
    ...rest
  },
  ref
) {
  return (
    <StyledButton
      {...rest}
      disabled={disabled}
      variant={variant}
      size={size}
      fullWidth={fixed}
      ref={ref}
    >
      {children}
    </StyledButton>
  )
})
export default Button

const StyledButton = styled(Clickable)
  .withConfig<StyledProps>({
    shouldForwardProp(prop) {
      // fixed は <button> 要素に渡ってはいけない
      return prop !== 'fullWidth'
    },
  })
  .attrs<StyledProps, ReturnType<typeof styledProps>>(styledProps)`
  width: ${(p) => (p.fullWidth ? 'stretch' : 'min-content')};
  display: inline-grid;
  align-items: center;
  justify-content: center;
  cursor: pointer;
  user-select: none;
  white-space: nowrap;

  ${(p) =>
    theme((o) => [
      o.font[p.font].hover.press,
      o.bg[p.background].hover.press,
      o.typography(14).bold.preserveHalfLeading,
      o.padding.horizontal(p.padding),
      o.disabled,
      o.borderRadius('oval'),
      o.outline.default.focus,
    ])}

  /* よく考えたらheight=32って定義が存在しないな... */
  height: ${(p) => p.height}px;
`

function styledProps(props: StyledProps) {
  return {
    ...props,
    ...variantToProps(props.variant),
    ...sizeToProps(props.size),
  }
}

function variantToProps(variant: Variant) {
  switch (variant) {
    case 'Overlay':
      return { font: 'text5', background: 'surface4' } as const
    case 'Default':
      return { font: 'text2', background: 'surface3' } as const
    case 'Primary':
      return { font: 'text5', background: 'brand' } as const
    case 'Navigation':
      return { font: 'text5', background: 'surface6' } as const
    case 'Danger':
      return { font: 'text5', background: 'assertive' } as const
    default:
      return unreachable(variant)
  }
}

function sizeToProps(size: Size) {
  switch (size) {
    case 'S':
      return {
        height: 32,
        padding: 16,
      } as const
    case 'M':
      return {
        height: 40,
        padding: 24,
      } as const
  }
}
