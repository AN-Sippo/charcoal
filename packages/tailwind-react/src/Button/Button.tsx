import classNames from 'classnames'
import { ForwardedRef, createElement, forwardRef } from 'react'

export type CustomJSXElement =
  | keyof JSX.IntrinsicElements
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  | React.JSXElementConstructor<any>

export type ButtonProps<T extends CustomJSXElement = 'button'> = {
  children?: React.ReactNode
  as?: T
  variant?: Variant
  size?: Size
  fullWidth?: boolean
} & Omit<React.ComponentProps<T>, 'children'>

type Variant = 'Primary' | 'Default' | 'Overlay' | 'Danger' | 'Navigation'
type Size = 'S' | 'M'

export const Button: <T extends CustomJSXElement = 'button'>(
  props: ButtonProps<T>
) => JSX.Element | null = forwardRef(function Button<
  T extends CustomJSXElement
>(
  {
    className,
    variant = 'Default',
    size = 'M',
    fullWidth,
    as: Component = 'button' as T,
    ...rest
  }: ButtonProps<T>,
  ref: ForwardedRef<ButtonProps<T>>
) {
  return createElement(Component, {
    ...rest,
    ref,
    className: classNames(
      'rounded-full flex cursor-pointer justify-center items-center border-0 font-[sans-serif] text-[14px] font-bold leading-[22px] transition-[box-shadow,background-color,color] duration-[0.2s] focus:shadow-focus disabled:opacity-[0.32] disabled:pointer-events-none',
      variant === 'Primary'
        ? 'bg-brand text-text5 active:bg-brand-press active:text-text5-press hover:bg-brand-hover hover:text-text5-hover'
        : variant === 'Overlay'
        ? 'bg-surface4 text-text5 active:bg-surface4-press active:text-text5-press hover:bg-surface4-hover hover:text-text5-hover'
        : variant === 'Navigation'
        ? 'bg-surface6 text-text5 active:bg-surface6-press active:text-text5-press hover:bg-surface6-hover hover:text-text5-hover'
        : variant === 'Danger'
        ? 'bg-assertive text-text5 active:bg-assertive-press active:text-text5-press hover:bg-assertive-hover hover:text-text5-hover'
        : 'bg-surface3 text-text2 active:bg-surface3-press active:text-text2-press hover:bg-surface3-hover hover:text-text2-hover',
      size === 'S' ? 'h-[32px] px-[16px]' : 'h-[40px] px-[24px]',
      fullWidth === true ? 'w-full' : 'w-fit',
      className
    ),
  })
})
