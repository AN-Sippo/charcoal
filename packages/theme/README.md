# `@charcoal/theme`

charcoal 全体で制約されるテーマ構造の型を提供します。

## インストール

npm

```
npm i --save-dev @charcoal/theme
```

yarn

```
yarn add -D @charcoal/theme
```
