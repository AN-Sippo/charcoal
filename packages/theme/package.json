{
  "name": "@charcoal-ui/theme",
  "version": "4.0.0-beta.15",
  "license": "Apache-2.0",
  "main": "./dist/index.cjs.js",
  "module": "./dist/index.esm.js",
  "exports": {
    "types": "./dist/index.d.ts",
    "require": "./dist/index.cjs.js",
    "import": "./dist/index.esm.js",
    "default": "./dist/index.esm.js"
  },
  "types": "./dist/index.d.ts",
  "sideEffects": [
    "*.css"
  ],
  "scripts": {
    "build": "npm-run-all --print-label --parallel 'build:*' --sequential serialize",
    "build:bundle": "FORCE_COLOR=1 tsup",
    "build:dts": "tsc --project tsconfig.build.json --pretty --emitDeclarationOnly",
    "build:copy-css": "cpx 'src/css/**/*.css' dist/css && cpx 'src/json/**/*.json' dist/json",
    "serialize": "node cli/index.js",
    "typecheck": "run-p --print-label 'typecheck:*'",
    "typecheck:main": "tsc --project tsconfig.build.json --pretty --noEmit",
    "typecheck:cli": "tsc --project cli/tsconfig.build.json --noEmit",
    "clean": "rimraf dist .tsbuildinfo",
    "test": "vitest run --passWithNoTests"
  },
  "devDependencies": {
    "cpx": "^1.5.0",
    "npm-run-all": "^4.1.5",
    "rimraf": "^3.0.2",
    "tsup": "^6.5.0",
    "typescript": "^4.9.5",
    "vitest": "^2.0.2"
  },
  "dependencies": {
    "@charcoal-ui/foundation": "^4.0.0-beta.15",
    "@charcoal-ui/utils": "^4.0.0-beta.15",
    "polished": "^4.1.4"
  },
  "files": [
    "src",
    "dist",
    "theme.json"
  ],
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "https://github.com/pixiv/charcoal.git",
    "directory": "packages/theme"
  },
  "gitHead": "e1ece2e43901ae667afdd5c178040607d939dcd5"
}
