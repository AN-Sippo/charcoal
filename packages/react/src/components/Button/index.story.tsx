import { useRef } from 'react'
import { ClickableElement } from '../Clickable'
import Button, { ButtonProps } from '.'
import { Meta, StoryObj } from '@storybook/react'

export default {
  title: 'Button',
  component: Button,
  argTypes: {
    variant: {
      control: {
        type: 'inline-radio',
        options: ['Primary', 'Default', 'Overlay', 'Danger', 'Navigation'],
      },
    },
    size: {
      control: {
        type: 'inline-radio',
        options: ['S', 'M'],
      },
    },
  },
  render: (args) => <Button {...args}>Button</Button>,
} as Meta<typeof Button>

export const Default: StoryObj<typeof Button> = {}

export const Primary: StoryObj<typeof Button> = {
  args: {
    variant: 'Primary',
  },
}

export const Navigation: StoryObj<typeof Button> = {
  args: {
    variant: 'Navigation',
  },
}

export const Overlay: StoryObj<typeof Button> = {
  args: {
    variant: 'Overlay',
  },
}

export const Danger: StoryObj<typeof Button> = {
  args: {
    variant: 'Danger',
  },
}

export const Small: StoryObj<typeof Button> = {
  args: {
    size: 'S',
  },
}

export const Fixed: StoryObj<typeof Button> = {
  args: {
    fullWidth: true,
  },
}

export const Disabled: StoryObj<typeof Button> = {
  args: {
    disabled: true,
  },
}

export const Active: StoryObj<typeof Button> = {
  args: {
    active: true,
  },
}

export const Link: StoryObj<typeof Button> = {
  args: {
    to: '#',
  },
}

export const Nihongo: StoryObj<typeof Button> = {
  render: (args) => <Button {...args}>日本語だよ</Button>,
}

export const Focus: StoryObj<typeof Button> = {
  render: function Render(args) {
    const ref = useRef<ClickableElement>(null)
    const focus = () => ref.current?.focus()
    return (
      <Button {...args} onMouseOver={focus} ref={ref}>
        Mouse over to focus
      </Button>
    )
  },
}

export const LayoutExample: StoryObj<typeof Button> = {
  render: (args: ButtonProps) => (
    <div
      css={`
        display: grid;
        gap: 24px;
      `}
    >
      <div
        css={`
          display: grid;
          grid: auto / auto-flow min-content;
          gap: 8px;
        `}
      >
        <Button {...args}>Grid</Button>
        <Button {...args} variant="Navigation">
          Layout
        </Button>
        <Button {...args} variant="Danger">
          Sample
        </Button>
      </div>
      <div
        css={`
          display: grid;
          grid: auto-flow auto / 392px;
          gap: 8px;
        `}
      >
        <Button {...args} variant="Primary" fullWidth>
          Submit
        </Button>
        <Button {...args} variant="Default" fullWidth>
          Cancel
        </Button>
      </div>
      <div
        css={`
          display: flex;

          & > * + * {
            margin-left: 8px;
          }
        `}
      >
        <Button {...args}>Flex</Button>
        <Button {...args} variant="Navigation">
          Layout
        </Button>
        <Button {...args} variant="Danger">
          Sample
        </Button>
      </div>
      <div
        css={`
          display: flex;
          flex-direction: column;
          width: 392px;

          & > * + * {
            margin-top: 8px;
          }
        `}
      >
        <Button {...args} variant="Primary" fullWidth>
          Submit
        </Button>
        <Button {...args} variant="Default" fullWidth>
          Cancel
        </Button>
      </div>
      <div
        css={`
          display: flex;
          flex-direction: column;
          width: 392px;

          & > * + * {
            margin-top: 8px;
          }
        `}
      >
        <Button {...args} variant="Primary" fullWidth>
          すべて見る
        </Button>
        <Button {...args} variant="Default">
          作品を投稿
        </Button>
      </div>
    </div>
  ),
}
