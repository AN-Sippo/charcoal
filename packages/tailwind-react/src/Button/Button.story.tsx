import { Button } from './Button'

import { Meta, StoryObj } from '@storybook/react'

const meta = {
  title: 'tailwind-react/Button',
  component: Button,
  args: {
    children: 'Button',
  },
  argTypes: {
    variant: {
      control: {
        type: 'inline-radio',
        options: ['Primary', 'Default', 'Overlay', 'Danger', 'Navigation'],
      },
    },
    size: {
      control: {
        type: 'inline-radio',
        options: ['S', 'M'],
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
  parameters: {
    controls: {
      exclude: /as/,
    },
  },
} satisfies Meta<typeof Button>

export default meta

export const Basic: StoryObj<typeof meta> = {
  args: { variant: 'Default' },
}
export const Primary: StoryObj<typeof meta> = {
  args: { variant: 'Primary' },
}
export const Overlay: StoryObj<typeof meta> = {
  args: { variant: 'Overlay' },
}
export const Danger: StoryObj<typeof meta> = {
  args: { variant: 'Danger' },
}
export const Navigation: StoryObj<typeof meta> = {
  args: { variant: 'Navigation' },
}
