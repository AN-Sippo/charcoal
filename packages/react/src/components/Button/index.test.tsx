import 'jest-styled-components'

import Button from '.'
import renderder from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { light } from '@charcoal-ui/theme'

export function render(children: JSX.Element) {
  return renderder
    .create(<ThemeProvider theme={light}>{children}</ThemeProvider>)
    .toJSON()
}

describe('Basic', () => {
  test('<Button>Hello</Button>', () => {
    expect(render(<Button>Hello</Button>)).toMatchSnapshot()
  })
})

describe('Link', () => {
  test('<Button to="#">Hello</Button>', () => {
    expect(render(<Button to="#">Hello</Button>)).toMatchSnapshot()
  })
})
