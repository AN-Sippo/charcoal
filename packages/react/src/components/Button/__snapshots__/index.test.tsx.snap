// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> componentAs props are passed to the data-as attribute 1`] = `
<a
  className="charcoal-button"
  data-as="componentAs"
  data-full-width={true}
  href="/"
  target="_blank"
/>
`;
